pmtiles/convert.go

```diff
@@ -117,6 +127,197 @@ func setZoomCenterDefaults(header *HeaderV3, entries []EntryV3) {
 	}
 }
 
+func convertDir(logger *log.Logger, input string, output string, deduplicate bool, tmpfile *os.File) error {
+	start := time.Now()
+
+	logger.Println("Pass 1: Assembling TileID set")
+
+	tileset := roaring64.New()
+	minZoom := uint8(255)
+	maxZoom := uint8(0)
+
+	minLat := int32(+180 * 1e7)
+	maxLat := int32(-180 * 1e7)
+	minLon := int32(+90 * 1e7)
+	maxLon := int32(-90 * 1e7)
+
+	tileFormat := ""
+
+	fileTypes := map[string]TileType{
+		"png":  Png,
+		"pbf":  Mvt,
+		"jpg":  Jpeg,
+		"webp": Webp,
+		"avif": Avif,
+	}
+
+	err := filepath.Walk(input, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		extension := strings.ToLower(filepath.Ext(info.Name()))[1:]
+
+		// skip files with unsupported extensions
+		if _, ok := fileTypes[extension]; !ok {
+			return nil
+		}
+
+		if tileFormat == "" {
+			tileFormat = extension
+		} else if tileFormat != extension {
+			return fmt.Errorf("mixed tile formats in directory: %s - enoountered both %s and %s", path, tileFormat, extension)
+		}
+
+		relPath, err := filepath.Rel(input, path)
+		if err != nil {
+			return err
+		}
+
+		parts := strings.Split(relPath, string(os.PathSeparator))
+		if len(parts) == 3 {
+			z, x, y := parts[0], parts[1], strings.TrimSuffix(parts[2], "."+extension)
+
+			zInt, err := strconv.ParseUint(z, 10, 8)
+			if err != nil {
+				return fmt.Errorf("invalid zoom level: %s", z)
+			}
+
+			if uint8(zInt) < minZoom {
+				minZoom = uint8(zInt)
+			}
+			if uint8(zInt) > maxZoom {
+				maxZoom = uint8(zInt)
+			}
+
+			xInt, err := strconv.ParseUint(x, 10, 32)
+			if err != nil {
+				return fmt.Errorf("invalid tile column: %s", x)
+			}
+
+			yInt, err := strconv.ParseUint(y, 10, 32)
+			if err != nil {
+				return fmt.Errorf("invalid tile row: %s", y)
+			}
+
+			_minLat, _minLon, _maxLat, _maxLon := TileToLatLonBounds(int(zInt), int(xInt), int(yInt))
```

Comment:
The tile coordinates are not enough to determine the intended geographic bounds for a tileset - tile pyramids that contain the z=0 tile will have a bounds of the entire web mercator extent. We need to read user-supplied bounds from a `metadata.json` or allow passing it on the command line.